File: c++/countPrimes.cpp
// leetcode.com 204. Count Primes

#include <vector>
#include <cmath>

using namespace std;

class Solution
{
  public:
    int countPrimes(int n)
    {
        vector<bool> s(n, true);

        int upper = sqrt(n);

        for (int i = 2; i <= upper; i++)
        {
            if (s[i])
            {
                for (int j = i * i; j < n; j += i)
                    s[j] = false;
            }
        }

        int res = 0;

        for (int j = 2; j < n; j++)
            if (s[j])
                res++;

        return res;
    }
};

File: leetcode/java/src/main/java/com/vkeonline/leetcode/p200/CountPrimes.java
package com.vkeonline.leetcode.p200;

import java.util.Arrays;

/**
 * Leetcode [E]: 204. Count Primes
 * @author csgear
 */
public class CountPrimes {
    public int countPrimes(int n) {
        boolean[] isPrime = new boolean[n];

        Arrays.fill(isPrime, true);

        int upper = (int) Math.sqrt(n);

        for (int i = 2; i <= upper; i++) {
            if (isPrime[i]) {
                for (int j = i * i; j < n; j += i)
                {
                    isPrime[j] = false;
                }
            }
        }

        int result = 0;

        for (boolean b : isPrime) {
            if (b) {
                result++;
            }
        }
        return result;
    }

}
